src/connection.ts

```diff
@@ -0,0 +1,457 @@
+import { Logger } from "./logger"
+import { CloseRequest } from "./requests/close_request"
+import { ExchangeCommandVersionsRequest } from "./requests/exchange_command_versions_request"
+import { OpenRequest } from "./requests/open_request"
+import { PeerPropertiesRequest } from "./requests/peer_properties_request"
+import { BufferSizeParams, BufferSizeSettings, Request } from "./requests/request"
+import { SaslAuthenticateRequest } from "./requests/sasl_authenticate_request"
+import { SaslHandshakeRequest } from "./requests/sasl_handshake_request"
+import { TuneRequest } from "./requests/tune_request"
+import {
+  ConsumerUpdateQueryListener,
+  DeliverListener,
+  MetadataUpdateListener,
+  PublishConfirmListener,
+  PublishErrorListener,
+  ResponseDecoder,
+} from "./response_decoder"
+import { CloseResponse } from "./responses/close_response"
+import { ExchangeCommandVersionsResponse } from "./responses/exchange_command_versions_response"
+import { OpenResponse } from "./responses/open_response"
+import { PeerPropertiesResponse } from "./responses/peer_properties_response"
+import { Response } from "./responses/response"
+import { SaslAuthenticateResponse } from "./responses/sasl_authenticate_response"
+import { SaslHandshakeResponse } from "./responses/sasl_handshake_response"
+import { TuneResponse } from "./responses/tune_response"
+import { DEFAULT_FRAME_MAX, DEFAULT_UNLIMITED_FRAME_MAX, removeFrom } from "./util"
+import { Version, checkServerDeclaredVersions, getClientSupportedVersions } from "./versions"
+import { WaitingResponse } from "./waiting_response"
+import { ClientListenersParams, ClientParams, ClosingParams, QueryOffsetParams, StoreOffsetParams } from "./client"
+import { QueryPublisherResponse } from "./responses/query_publisher_response"
+import { QueryPublisherRequest } from "./requests/query_publisher_request"
+import { StoreOffsetRequest } from "./requests/store_offset_request"
+import { QueryOffsetResponse } from "./responses/query_offset_response"
+import { QueryOffsetRequest } from "./requests/query_offset_request"
+
+export type ConnectionClosedListener = (hadError: boolean) => void
+
+export type ConnectionProxyListenersParams = ClientListenersParams & {
+  deliver?: DeliverListener
+  consumer_update_query?: ConsumerUpdateQueryListener
+}
+
+export type ConnectionProxyParams = ClientParams & {
+  listeners?: ConnectionProxyListenersParams
+}
+
+export type ConnectionInfo = {
+  host: string
+  port: number
+  id: string
+  readable?: boolean
+  writable?: boolean
+  localPort?: number
+}
+
+function extractHeartbeatInterval(heartbeatInterval: number, tuneResponse: TuneResponse): number {
+  return heartbeatInterval === 0 ? tuneResponse.heartbeat : Math.min(heartbeatInterval, tuneResponse.heartbeat)
+}
+
+export class Connection {
+  public readonly hostname: string
+  public readonly leader: boolean
+  public readonly streamName: string | undefined
+  private socket: Socket
+  private correlationId = 100
```

Comment:
I know it's not relevant to the PR (it was already there) but do we know why we have this default?

Reply:
I'm not sure. @albertobarrila do you have some insight on this? Anyway, I think it would be better to address this in another issue, if we decide to change it